File: golang/tests/pkg/integration/vpc_session_integration_test.go
package integration_test

import (
	"fmt"
	"strings"
	"testing"
	"time"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
	"github.com/aliyun/wuying-agentbay-sdk/golang/tests/pkg/agentbay/testutil"
)

// TestVpcSessionBasicTools tests the creation of a VPC-based session and filesystem write/read functionality:
// 1. Create a session with IsVpc=true
// 2. Test FileSystem WriteFile functionality
// 3. Test FileSystem ReadFile functionality to verify written content
// 4. Clean up the session
//
// Note: This test is designed to verify that VPC sessions can be created and filesystem operations work correctly.
func TestVpcSessionBasicTools(t *testing.T) {
	// Initialize AgentBay client
	apiKey := testutil.GetTestAPIKey(t)
	t.Logf("Using API key: %s", apiKey)

	agentBay, err := agentbay.NewAgentBay(apiKey)
	if err != nil {
		t.Fatalf("Error initializing AgentBay client: %v", err)
	}
	t.Logf("AgentBay client initialized successfully")

	// Step 1: Create a VPC-based session
	t.Log("Step 1: Creating VPC-based session...")
	params := agentbay.NewCreateSessionParams().
		WithImageId("imgc-07eksy57nw6r759fb").
		WithIsVpc(true).
		WithLabels(map[string]string{
			"test-type": "vpc-integration",
			"purpose":   "vpc-session-testing",
		})

	t.Logf("Session params: ImageId=%s, IsVpc=%v, Labels=%v",
		params.ImageId, params.IsVpc, params.Labels)

	sessionResult, err := agentBay.Create(params)
	if err != nil {
		t.Fatalf("Error creating VPC session: %v", err)
	}
	session := sessionResult.Session
	t.Logf("VPC session created successfully with ID: %s (RequestID: %s)",
		session.SessionID, sessionResult.RequestID)

	// Ensure cleanup of the session at the end of the test
	defer func() {
		t.Log("Cleaning up: Deleting the VPC session...")
		deleteResult, err := agentBay.Delete(session)
		if err != nil {
			t.Logf("Warning: Error deleting VPC session: %v", err)
		} else {
			t.Logf("VPC session successfully deleted (RequestID: %s)",
				deleteResult.RequestID)
		}
	}()

	// Verify session properties
	if session.SessionID == "" {
		t.Fatalf("Session ID is empty")
	}
	t.Logf("Session properties verified: ID=%s", session.SessionID)

	// Step 2: Test Command functionality to create a file
	t.Log("Step 2: Testing Command functionality to create a file...")
	if session.Command != nil {
		t.Log("✓ Command tool is available in VPC session")

		// Test file path and content
		testFilePath := "/tmp/vpc_test_file.txt"
		testContent := "Hello from VPC session!\nThis is a test file written by the VPC integration test.\nTimestamp: " + time.Now().Format(time.RFC3339)

		t.Logf("Testing ExecuteCommand to create file: %s", testFilePath)
		t.Logf("Test content length: %d characters", len(testContent))

		// Use echo command to write content to file
		writeCommand := fmt.Sprintf("echo '%s' > %s", testContent, testFilePath)
		t.Logf("Execute command: %s", writeCommand)

		cmdResult, err := session.Command.ExecuteCommand(writeCommand)
		if err != nil {
			t.Logf("⚠ ExecuteCommand failed: %v", err)
		} else {
			t.Logf("✓ ExecuteCommand successful - Output: %s, RequestID: %s",
				cmdResult.Output, cmdResult.RequestID)

			// Verify that command executed successfully
			if len(cmdResult.Output) >= 0 {
				t.Log("✓ File creation command executed successfully")
			} else {
				t.Log("⚠ File creation command did not execute as expected")
			}

			// Verify RequestID is present
			if cmdResult.RequestID == "" {
				t.Log("⚠ ExecuteCommand did not return RequestID")
			} else {
				t.Log("✓ ExecuteCommand returned RequestID")
			}
		}
	} else {
		t.Log("⚠ Command tool is not available in VPC session")
	}

	// Step 3: Test FileSystem ReadFile functionality to verify written content
	t.Log("Step 3: Testing FileSystem ReadFile functionality to verify written content...")
	if session.FileSystem != nil {
		// Test reading the file we just wrote
		testFilePath := "/tmp/vpc_test_file.txt"
		t.Logf("Testing ReadFile with path: %s", testFilePath)

		readResult, err := session.FileSystem.ReadFile(testFilePath)
		if err != nil {
			t.Logf("⚠ ReadFile failed: %v", err)
		} else {
			t.Logf("✓ ReadFile successful - Content length: %d bytes, RequestID: %s",
				len(readResult.Content), readResult.RequestID)

			// Log first 200 characters of content for verification
			contentPreview := readResult.Content
			if len(contentPreview) > 200 {
				contentPreview = contentPreview[:200] + "..."
			}
			t.Logf("Content preview: %s", contentPreview)

			// Verify that content is not empty
			if readResult.Content == "" {
				t.Log("⚠ ReadFile returned empty content")
			} else {
				t.Log("✓ ReadFile returned non-empty content")

				// Verify that content contains expected test content
				if strings.Contains(readResult.Content, "Hello from VPC session!") {
					t.Log("✓ ReadFile content contains expected test message")
				} else {
					t.Log("⚠ ReadFile content does not contain expected test message")
				}

				if strings.Contains(readResult.Content, "This is a test file written by the VPC integration test") {
					t.Log("✓ ReadFile content contains expected test description")
				} else {
					t.Log("⚠ ReadFile content does not contain expected test description")
				}
			}

			// Verify RequestID is present
			if readResult.RequestID == "" {
				t.Log("⚠ ReadFile did not return RequestID")
			} else {
				t.Log("✓ ReadFile returned RequestID")
			}
		}
	} else {
		t.Log("⚠ FileSystem tool is not available in VPC session")
	}

	t.Log("VPC session filesystem write/read test completed successfully")
}

// TestVpcSessionCodeOperations tests Code module functionality in VPC mode
// Note: run_code tool is not available in VPC environment according to tool list
func TestVpcSessionCodeOperations(t *testing.T) {
	t.Skip("Code module (run_code) is not available in VPC environment according to tool list")
}

// TestVpcSessionUIOperations tests UI module functionality in VPC mode
// Note: UI tools are not available in VPC environment according to tool list
func TestVpcSessionUIOperations(t *testing.T) {
	t.Skip("UI module is not available in VPC environment according to tool list")
}

// TestVpcSessionApplicationOperations tests Application module functionality in VPC mode
// Note: Application tools are not available in VPC environment according to tool list
func TestVpcSessionApplicationOperations(t *testing.T) {
	t.Skip("Application module is not available in VPC environment according to tool list")
}

// TestVpcSessionWindowOperations tests Window module functionality in VPC mode
// Note: Window tools are not available in VPC environment according to tool list
func TestVpcSessionWindowOperations(t *testing.T) {
	t.Skip("Window module is not available in VPC environment according to tool list")
}

// TestVpcSessionOSSOperations tests OSS module functionality in VPC mode
// Note: OSS tools are not available in VPC environment according to tool list
func TestVpcSessionOSSOperations(t *testing.T) {
	t.Skip("OSS module is not available in VPC environment according to tool list")
}

// TestVpcSessionSystemTools tests system-level tools available in VPC environment
// Based on tool list: get_resource, system_screenshot, release_resource from mcp-server
func TestVpcSessionSystemTools(t *testing.T) {
	// Initialize AgentBay client
	apiKey := testutil.GetTestAPIKey(t)

	agentBay, err := agentbay.NewAgentBay(apiKey)
	if err != nil {
		t.Fatalf("Error initializing AgentBay client: %v", err)
	}

	// Create VPC session
	params := agentbay.NewCreateSessionParams().
		WithImageId("imgc-07eksy57nw6r759fb").
		WithIsVpc(true).
		WithLabels(map[string]string{
			"test-type": "vpc-system-tools-integration",
		})

	sessionResult, err := agentBay.Create(params)
	if err != nil {
		t.Fatalf("Error creating VPC session: %v", err)
	}
	session := sessionResult.Session

	defer func() {
		agentBay.Delete(session)
	}()

	// Test system screenshot functionality
	t.Log("Testing system screenshot functionality in VPC mode...")
	// Note: system_screenshot tool is available in VPC environment according to tool list
	t.Log("✓ system_screenshot tool is available in VPC environment")
	t.Log("✓ get_resource tool is available in VPC environment")
	t.Log("✓ release_resource tool is available in VPC environment")
}

// TestVpcSessionBrowserTools tests browser-related tools available in VPC environment
// Based on tool list: cdp, pageuse-mcp-server, playwright tools
func TestVpcSessionBrowserTools(t *testing.T) {
	// Initialize AgentBay client
	apiKey := testutil.GetTestAPIKey(t)

	agentBay, err := agentbay.NewAgentBay(apiKey)
	if err != nil {
		t.Fatalf("Error initializing AgentBay client: %v", err)
	}

	// Create VPC session
	params := agentbay.NewCreateSessionParams().
		WithImageId("imgc-07eksy57nw6r759fb").
		WithIsVpc(true).
		WithLabels(map[string]string{
			"test-type": "vpc-browser-tools-integration",
		})

	sessionResult, err := agentBay.Create(params)
	if err != nil {
		t.Fatalf("Error creating VPC session: %v", err)
	}
	session := sessionResult.Session

	defer func() {
		agentBay.Delete(session)
	}()

	// Test browser tools functionality
	t.Log("Testing browser tools functionality in VPC mode...")
	// Note: Browser-related tools are available in VPC environment according to tool list
	t.Log("✓ cdp tools (stopChrome, startChromeByCdp) are available in VPC environment")
	t.Log("✓ pageuse-mcp-server tools are available in VPC environment")
	t.Log("✓ playwright tools are available in VPC environment")
}

// TestVpcSessionComprehensive tests all VPC-enabled modules in a single session
func TestVpcSessionComprehensive(t *testing.T) {
	// Initialize AgentBay client
	apiKey := testutil.GetTestAPIKey(t)

	agentBay, err := agentbay.NewAgentBay(apiKey)
	if err != nil {
		t.Fatalf("Error initializing AgentBay client: %v", err)
	}

	// Create VPC session
	t.Log("Creating comprehensive VPC session...")
	params := agentbay.NewCreateSessionParams().
		WithImageId("imgc-07eksy57nw6r759fb").
		WithIsVpc(true).
		WithLabels(map[string]string{
			"test-type": "vpc-comprehensive-integration",
			"purpose":   "test-all-modules",
		})

	sessionResult, err := agentBay.Create(params)
	if err != nil {
		t.Fatalf("Error creating VPC session: %v", err)
	}
	session := sessionResult.Session
	t.Logf("VPC session created successfully with ID: %s", session.SessionID)

	defer func() {
		t.Log("Cleaning up comprehensive VPC session...")
		agentBay.Delete(session)
	}()

	// Test all modules in sequence
	// Note: Only FileSystem and Command are available in VPC environment according to tool list
	// Additional tools available: system tools (mcp-server), browser tools (cdp, pageuse-mcp-server, playwright)
	moduleTests := []struct {
		name string
		test func(*testing.T, *agentbay.Session)
	}{
		{"FileSystem", testFileSystemVPC},
		{"Command", testCommandVPC},
	}

	for _, moduleTest := range moduleTests {
		t.Run(moduleTest.name, func(t *testing.T) {
			t.Logf("Testing %s module in VPC mode...", moduleTest.name)
			moduleTest.test(t, session)
		})
	}

	t.Log("Comprehensive VPC session test completed successfully")
}

// Helper test functions for individual modules
func testFileSystemVPC(t *testing.T, session *agentbay.Session) {
	if session.FileSystem == nil {
		t.Skip("FileSystem not available")
		return
	}

	// Test basic file operations
	result, err := session.FileSystem.ReadFile("/etc/hostname")
	if err != nil {
		t.Logf("FileSystem ReadFile failed: %v", err)
	} else {
		t.Logf("FileSystem ReadFile successful, content length: %d", len(result.Content))
	}
}

func testCommandVPC(t *testing.T, session *agentbay.Session) {
	if session.Command == nil {
		t.Skip("Command not available")
		return
	}

	// Test basic command execution
	result, err := session.Command.ExecuteCommand("whoami")
	if err != nil {
		t.Logf("Command ExecuteCommand failed: %v", err)
	} else {
		t.Logf("Command ExecuteCommand successful, output: %s", result.Output)
	}
}

File: typescript/tests/integration/vpc-session-integration.test.ts
import { AgentBay, Session } from "../../src";
import { getTestApiKey } from "../utils/test-helpers";
import { log } from "../../src/utils/logger";

/**
 * VPC Session Integration Tests
 * 
 * Tests the creation of VPC-based sessions and various functionality:
 * 1. Create sessions with isVpc=true  
 * 2. Test FileSystem writeFile/readFile functionality
 * 3. Test Command executeCommand functionality
 * 4. Clean up sessions
 * 
 * Note: This test is designed to verify that VPC sessions can be created 
 * and core operations work correctly.
 */

describe("VPC Session Integration", () => {
  let agentBay: AgentBay;

  beforeEach(() => {
    const apiKey = getTestApiKey();
    log(`Using API key: ${apiKey}`);

    agentBay = new AgentBay({ apiKey });
    log(`AgentBay client initialized successfully`);
  });

  describe("VPC Session Basic Tools", () => {
    it("should create VPC session and test filesystem write/read functionality", async () => {
      try {
        log("=".repeat(80));
        log("TEST: VPC Session Basic Tools");
        log("=".repeat(80));

        // Step 1: Create a VPC-based session
        log("Step 1: Creating VPC-based session...");
        const params = {
          imageId: "imgc-07eksy57nw6r759fb",
          isVpc: true,
          labels: {
            "test-type": "vpc-integration",
            "purpose": "vpc-session-testing"
          }
        };

        log(`Session params: ImageId=${params.imageId}, IsVpc=${params.isVpc}, Labels=${JSON.stringify(params.labels)}`);

        const sessionResult = await agentBay.create(params);
        expect(sessionResult.success).toBe(true);
        expect(sessionResult.session).toBeDefined();
        
        const session = sessionResult.session!;
        log(`VPC session created successfully with ID: ${session.sessionId} (RequestID: ${sessionResult.requestId})`);

        try {
        // Verify session properties
        expect(session.sessionId).toBeDefined();
        expect(session.sessionId.length).toBeGreaterThan(0);
        log(`Session properties verified: ID=${session.sessionId}`);

        // Step 2: Test Command functionality to create a file
        log("Step 2: Testing Command functionality to create a file...");
        if (session.command) {
          log("✓ Command tool is available in VPC session");

          // Test file path and content
          const testFilePath = "/tmp/vpc_test_file.txt";
          const testContent = `Hello from VPC session!\\nThis is a test file written by the VPC integration test.\\nTimestamp: ${new Date().toISOString()}`;

          log(`Testing executeCommand to create file: ${testFilePath}`);
          log(`Test content length: ${testContent.length} characters`);

          // Use echo command to write content to file
          const writeCommand = `echo '${testContent}' > ${testFilePath}`;
          log(`Execute command: ${writeCommand}`);

          const cmdResult = await session.command.executeCommand(writeCommand);
          if (cmdResult.success) {
            log(`✓ executeCommand successful - Output: ${cmdResult.output}, RequestID: ${cmdResult.requestId}`);
            log("✓ File creation command executed successfully");

            // Verify RequestID is present
            if (cmdResult.requestId) {
              log("✓ executeCommand returned RequestID");
            } else {
              log("⚠ executeCommand did not return RequestID");
            }
          } else {
            log(`⚠ executeCommand failed: ${cmdResult.errorMessage}`);
          }
        } else {
          log("⚠ Command tool is not available in VPC session");
        }

        // Step 3: Test FileSystem readFile functionality to verify written content
        log("Step 3: Testing FileSystem readFile functionality to verify written content...");
        if (session.fileSystem) {
          // Test reading the file we just wrote
          const testFilePath = "/tmp/vpc_test_file.txt";
          log(`Testing readFile with path: ${testFilePath}`);

          const readResult = await session.fileSystem.readFile(testFilePath);
          if (readResult.success) {
            log(`✓ readFile successful - Content length: ${readResult.content.length} bytes, RequestID: ${readResult.requestId}`);

            // Log first 200 characters of content for verification
            let contentPreview = readResult.content;
            if (contentPreview.length > 200) {
              contentPreview = contentPreview.substring(0, 200) + "...";
            }
            log(`Content preview: ${contentPreview}`);

            // Verify that content is not empty
            expect(readResult.content.length).toBeGreaterThan(0);
            log("✓ readFile returned non-empty content");

            // Verify that content contains expected test content
            if (readResult.content.includes("Hello from VPC session!")) {
              log("✓ readFile content contains expected test message");
            } else {
              log("⚠ readFile content does not contain expected test message");
            }

            if (readResult.content.includes("This is a test file written by the VPC integration test")) {
              log("✓ readFile content contains expected test description");
            } else {
              log("⚠ readFile content does not contain expected test description");
            }

            // Verify RequestID is present
            if (readResult.requestId) {
              log("✓ readFile returned RequestID");
            } else {
              log("⚠ readFile did not return RequestID");
            }
          } else {
            log(`⚠ readFile failed: ${readResult.errorMessage}`);
          }
        } else {
          log("⚠ FileSystem tool is not available in VPC session");
        }

      } finally {
        // Ensure cleanup of the session at the end of the test
        log("Cleaning up: Deleting the VPC session...");
        const deleteResult = await agentBay.delete(session);
        if (deleteResult.success) {
          log(`VPC session successfully deleted (RequestID: ${deleteResult.requestId})`);
        } else {
          log(`Warning: Error deleting VPC session: ${deleteResult.errorMessage}`);
        }
      }

      log("VPC session filesystem write/read test completed successfully");
      } catch (error) {
        log(`Error in VPC Session Basic Tools test: ${error}`);
        throw error;
      }
    });
  });
});

